test/k8sT/DatapathConfiguration.go

```diff
@@ -318,10 +323,14 @@ var _ = Describe("K8sDatapathConfig", func() {
 				"tunnel":                 "disabled",
 				"k8s.requireIPv4PodCIDR": "true",
 				"endpointRoutes.enabled": "true",
-			}, DeployCiliumOptionsAndDNS)
+			}, RedeployCiliumAfterDatapathChange)
 
 			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue(), "Connectivity test between nodes failed")
 		})
+
+		// CAUTION: endpointRoutes mode was toggled. The next test
+		// MUST redeploy kubedns via RedeployCiliumAfterDatapathChange
+		// or potentially face GH-16717.
```

Comment:
Couldn't we have a variable in `kubectl` to remember the current state of endpoint routes and use that to detect when we should restart DNS pods?

Reply:
I briefly thought about this but discounted it with some hand-waving about storing state across tests being bad. Coming fresh after the weekend, I think this would be a much cleaner approach, so I'll try this out.

Reply:
> storing state across tests being bad

Yeah, I was concerned about that as well, but relying on people to call the correct code in the correct order is probably worse.